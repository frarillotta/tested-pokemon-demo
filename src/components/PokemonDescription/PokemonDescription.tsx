import Image from "next/image";
import { useQuery } from "react-query";
import { PokeballButton } from "../PokeballButton/PokeballButton";
import { PokemonMainInfo } from "../PokemonMainInfo/PokemonMainInfo";
import { PokemonStatsInfo } from "../PokemonStatsInfo/PokemonStatsInfo";
import { apiClient } from "../../API/apiClient";
import { Pokemon, pokemonResults } from "../../types";
import styled from "styled-components";
import { memo, useState } from "react";
import AppBar from '@material-ui/core/AppBar';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import Box from "@material-ui/core/Box";
import {getColorByType, getPokemonId} from "../../utils";
import {PokemonMovesInfo} from "../PokemonMovesInfo/PokemonMovesInfo";

type PokemonDescriptionProps = {
	activePokemon: pokemonResults,
	capturedPokemons: Record<string, string>,
	addPokemon: (pokemon: pokemonResults) => void,
	removePokemon: (pokemonName: string) => void,
	className: string,
}

export const PokemonDescription = memo(({activePokemon, capturedPokemons, addPokemon, removePokemon, className}: PokemonDescriptionProps) => {

	const [value, setValue] = useState(0);

	const handleChange = (_, newValue) => {
	  setValue(newValue);
	};

	const fetchPokemons = async () => apiClient(activePokemon.url);

	const {
		data,
		error,
		isError,
		isFetching,
		isIdle,
		isLoading,
		isSuccess,
	} = useQuery<Pokemon>(activePokemon.name, fetchPokemons, {});
	const pokemon = isSuccess && data;
	const types = isSuccess && data && pokemon.types.map((type) => type.type.name);
	const isCaptured = pokemon && capturedPokemons.hasOwnProperty(pokemon.name);
	return (isSuccess && data && 
		<DescriptionWrapper backgroundColor={ getColorByType(types[0]) } className={className}>
			<Header>
				<PokemonImageWrapper>
					<Image 
						src={pokemon.sprites.front_default}
						alt={`${pokemon.name} front sprite`}
						width={200} 
						height={200}
					/>
				</PokemonImageWrapper>
				<PokemonTypesNameWrapper>
					<PokemonName>{pokemon.name.toUpperCase()} <PokemonId>#{getPokemonId(pokemon.id)}</PokemonId></PokemonName>
					<PokemonTypesWrapper>
						{types.map((type) => <PokemonType background={ getColorByType(type) } key={type}>{type}</PokemonType>)}
					</PokemonTypesWrapper>
				</PokemonTypesNameWrapper>
				<Pokeball
					isCaptured={isCaptured}
					onClick={() => isCaptured ? removePokemon(pokemon.name) : addPokemon(activePokemon)}
					size={100}
				/>
			</Header>
			<Wrapper>
				<AppBar 
					color="transparent"
					position="static"
					style={{
						boxShadow: "none"
					}}
				>
					<Tabs
						value={value} 
						onChange={handleChange}
						aria-label="Pokemon description tabs" 
						centered
					>
						<Tab label="Info" />
						<Tab label="Stats" />
						<Tab label="Moves" />
					</Tabs>
				</AppBar>
				<TabPanel value={value} index={0}>
					<PokemonMainInfo 
						abilities={pokemon.abilities} 
						weight={pokemon.weight} 
						height={pokemon.height}
					/>
				</TabPanel>
				<TabPanel value={value} index={1}>
					<PokemonStatsInfo stats={pokemon.stats}/>
				</TabPanel>
				<TabPanel value={value} index={2}>
					<PokemonMovesInfo moves={pokemon.moves}/>
				</TabPanel>
			</Wrapper>
		</DescriptionWrapper>
	)

//TODO stats with progress bar, map colors to stat name
//Tab for about / moves / stats?
//idk if image slider for different sprites
});

function TabPanel({ children, value, index, ...other }) {
  
	return (
	  <div
			role="tabpanel"
			hidden={value !== index}
			id={`simple-tabpanel-${index}`}
			aria-labelledby={`simple-tab-${index}`}
			{...other}
		>
			{value === index && (
				<Box p={3}>
					{children}
				</Box>
			)}
	  </div>
	);
}

const Header = styled.header`
`

const DescriptionWrapper = styled.main`
	position: relative;
	border-radius: 50px;
	margin: 20px 20px 20px 0px;
	background: ${props => props.backgroundColor};
`

const PokemonId = styled.span`
	margin-left: 15px;
	font-weight: normal;
	font-size: 0.9rem;
`

const Pokeball = styled(PokeballButton)`
	position: absolute;
	right: 17px;
	top: 20px;
`

const PokemonTypesNameWrapper = styled.div`
	position: absolute;
	top: 30px;
	left: 50px;
	display: flex;
	flex-direction: column;
	gap: 30px;
	@media(max-width: 770px) {		
		top: 25px;
		gap: 10px;
	}
`

const PokemonTypesWrapper = styled.div`
	display: flex;
	gap: 10px;
	@media(max-width: 770px) {		
		flex-direction: column;
	}
`;

const Wrapper = styled.div`
	background: white;
	padding-top: 50px;
	border-radius: 50px;
	margin-top: 135px;
	box-shadow: 2px 5px 18px 0px black;
`

const PokemonImageWrapper = styled.div`
	position: absolute;
	top: 0px;
	left: calc(50% - 100px);
`

const PokemonName = styled.div`
	font-weight: bold;
	letter-spacing: 2px;
	color: white;
`;

const PokemonType = styled.span`
	background: ${props => props.background};
	letter-spacing: 2px;
	color: white;
	border-radius: 16px;
	padding: 10px;
	width: fit-content;
	font-size: .8rem;
	font-weight: bold;
	box-shadow: 0px 1px 3px 0px black;
`;