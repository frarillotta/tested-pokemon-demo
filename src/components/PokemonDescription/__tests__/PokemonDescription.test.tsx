import { render } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "react-query";
import {PokemonDescription} from "../PokemonDescription";
import metapod from "../../../../__mocks__/dataMocks/metapod.json"
beforeAll(() => jest.spyOn(window, 'fetch'));

jest.mock("next/image", () => {
	// eslint-disable-next-line react/display-name
	return () => <></>;
});
describe("PokemonDescription", () => {
	const activePokemon = {
		name: "metapod",
		url: "https://www.mockpokemonUrl.pk/metapod"
	}
	const queryClient = new QueryClient();
	describe("when it has a captured pokemon match", () => {

		const capturedPokemons = {
			"metapod": "https://www.mockpokemonUrl.pk"
		}
		
		it("should have a consistent snapshot", () => {
			//@ts-ignore
			window.fetch.mockResolvedValueOnce({
				ok: true,
				json: async () => (metapod),
			})
			const {container} = render(<QueryClientProvider client={queryClient}>
				<PokemonDescription 
					activePokemon={activePokemon} 
					addPokemon={jest.fn()} 
					removePokemon={jest.fn()} 
					capturedPokemons={capturedPokemons}
				/>
			</QueryClientProvider>)
			expect(container).toMatchSnapshot();
		});

	});

	describe("when it doesnt have captured pokemon match", () => {

		const capturedPokemons = {
			"togepi": "https://www.mockpokemonUrl.pk"
		}

		it("should have a consistent snapshot", () => {
			
			//@ts-ignore
			window.fetch.mockResolvedValueOnce({
				ok: true,
				json: async () => (metapod),
			})
			const {container} = render(<QueryClientProvider client={queryClient}><PokemonDescription 
				activePokemon={activePokemon} 
				addPokemon={jest.fn()} 
				removePokemon={jest.fn()} 
				capturedPokemons={capturedPokemons}
			/>
			</QueryClientProvider>)
			expect(container).toMatchSnapshot();
		})
	})
	
})