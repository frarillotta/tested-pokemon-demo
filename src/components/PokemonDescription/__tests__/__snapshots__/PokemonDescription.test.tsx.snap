// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonDescription when it doesnt have captured pokemon match should have a consistent snapshot 1`] = `
<div>
  <main
    class="PokemonDescription__DescriptionWrapper-sc-rkyphs-1 evrziB"
  >
    <header
      class="PokemonDescription__Header-sc-rkyphs-0"
    >
      <div
        class="PokemonDescription__PokemonImageWrapper-sc-rkyphs-7 dnAJzR"
      />
      <div
        class="PokemonDescription__PokemonTypesNameWrapper-sc-rkyphs-4 gbgilG"
      >
        <div
          class="PokemonDescription__PokemonName-sc-rkyphs-8 dGaXHa"
        >
          SQUIRTLE
           
          <span
            class="PokemonDescription__PokemonId-sc-rkyphs-2 hReLpg"
          >
            #
            007
          </span>
        </div>
        <div
          class="PokemonDescription__PokemonTypesWrapper-sc-rkyphs-5 dCzXyk"
        >
          <span
            class="PokemonDescription__PokemonType-sc-rkyphs-9 iIyWrb"
          >
            water
          </span>
        </div>
      </div>
      <div
        class="PokeballButton__PokeballWrapper-sc-z0qx2i-0 jUctph PokemonDescription__Pokeball-sc-rkyphs-3 hpPmap"
        data-t-is-captured="false"
        data-t-pokeball="description-squirtle"
      >
        <svg
          class="PokeballButton__PokeballSVG-sc-z0qx2i-1 fWWEza"
          height="100"
          viewBox="25 25 50 50"
          width="100"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M 30 50
 a 1 1 1 0 1 40 0
 h-12.5
 a 1 1 1 0 0 -15 0
 z"
            fill="#f00"
            stroke="#222"
          />
          <circle
            cx="50"
            cy="50"
            fill="#222"
            r="5"
            stroke="#222"
          />
          <path
            d="M 30 50
 a 1 1 1 0 0 40 0
 h-12.5
 a 1 1 1 0 1 -15 0
 z"
            fill="#fff"
            stroke="#222"
          />
        </svg>
      </div>
    </header>
    <div
      class="PokemonDescription__Wrapper-sc-rkyphs-6 bqONiD"
    >
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorTransparent MuiPaper-elevation4"
        style="box-shadow: none;"
      >
        <div
          class="MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed"
            style="overflow: hidden;"
          >
            <div
              aria-label="Pokemon description tabs"
              class="MuiTabs-flexContainer MuiTabs-centered"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit Mui-selected"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Info
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Stats
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  Moves
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="PrivateTabIndicator-root-6 PrivateTabIndicator-colorSecondary-8 MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </header>
      <div
        aria-labelledby="simple-tab-0"
        id="simple-tabpanel-0"
        role="tabpanel"
      >
        <div
          class="MuiBox-root MuiBox-root-10"
        >
          <div
            class="PokemonMainInfo__PokemonInfo-sc-1y1zaw-3 eFwIDo"
          >
            <div
              class="PokemonMainInfo__InfoPropertyValueWrapper-sc-1y1zaw-0 jVYAZk"
            >
              <span
                class="PokemonMainInfo__InfoProperty-sc-1y1zaw-1 gtLCOd"
              >
                Height:
              </span>
              <span
                class="PokemonMainInfo__InfoValue-sc-1y1zaw-2 hqHXIZ"
              >
                50 cm ~(1'8'')
              </span>
            </div>
            <div
              class="PokemonMainInfo__InfoPropertyValueWrapper-sc-1y1zaw-0 jVYAZk"
            >
              <span
                class="PokemonMainInfo__InfoProperty-sc-1y1zaw-1 gtLCOd"
              >
                Weight:
              </span>
              <span
                class="PokemonMainInfo__InfoValue-sc-1y1zaw-2 hqHXIZ"
              >
                9 kg (19.85 lbs)
              </span>
            </div>
            <div
              class="PokemonMainInfo__InfoPropertyValueWrapper-sc-1y1zaw-0 jVYAZk"
            >
              <span
                class="PokemonMainInfo__InfoProperty-sc-1y1zaw-1 gtLCOd"
              >
                Abilities:
              </span>
              <span
                class="PokemonMainInfo__InfoValue-sc-1y1zaw-2 hqHXIZ"
              >
                Torrent, Rain dish
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-labelledby="simple-tab-1"
        hidden=""
        id="simple-tabpanel-1"
        role="tabpanel"
      />
      <div
        aria-labelledby="simple-tab-2"
        hidden=""
        id="simple-tabpanel-2"
        role="tabpanel"
      />
    </div>
  </main>
</div>
`;

exports[`PokemonDescription when it has a captured pokemon match should have a consistent snapshot 1`] = `<div />`;
